File: Code-Samples/Python/02-Text-and-Language/translate_text.py
#!/usr/bin/env python3
"""
Azure AI Translator - Text Translation Example

This example demonstrates how to use Azure AI Translator service to:
- Detect language of input text
- Translate text to multiple target languages
- Handle various translation scenarios
- Implement best practices for production use

Prerequisites:
- Azure subscription with Translator resource
- pip install requests python-dotenv

Set environment variables:
- TRANSLATOR_KEY: Your Translator resource key
- TRANSLATOR_ENDPOINT: Your Translator resource endpoint
- TRANSLATOR_REGION: Your Translator resource region
"""

import os
import json
import requests
import uuid
from typing import List, Dict, Optional, Union
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class AzureTranslator:
    """Azure AI Translator client for text translation operations."""
    
    def __init__(self):
        """Initialize the translator client with Azure credentials."""
        self.key = os.getenv('TRANSLATOR_KEY')
        self.endpoint = os.getenv('TRANSLATOR_ENDPOINT')
        self.region = os.getenv('TRANSLATOR_REGION')
        
        if not all([self.key, self.endpoint, self.region]):
            raise ValueError("Missing required environment variables. Please set TRANSLATOR_KEY, TRANSLATOR_ENDPOINT, and TRANSLATOR_REGION")
        
        # Ensure endpoint ends with /
        if not self.endpoint.endswith('/'):
            self.endpoint += '/'
            
        self.session = requests.Session()
        self.session.headers.update({
            'Ocp-Apim-Subscription-Key': self.key,
            'Ocp-Apim-Subscription-Region': self.region,
            'Content-type': 'application/json',
            'X-ClientTraceId': str(uuid.uuid4())
        })
    
    def detect_language(self, texts: Union[str, List[str]]) -> List[Dict]:
        """
        Detect the language of input text(s).
        
        Args:
            texts: String or list of strings to detect language for
            
        Returns:
            List of detection results with language, confidence, and alternatives
        """
        if isinstance(texts, str):
            texts = [texts]
        
        body = [{'text': text} for text in texts]
        
        try:
            response = self.session.post(
                f"{self.endpoint}detect?api-version=3.0",
                json=body
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error detecting language: {e}")
            return []
    
    def translate_text(self, 
                      texts: Union[str, List[str]], 
                      target_languages: Union[str, List[str]],
                      source_language: Optional[str] = None,
                      text_type: str = 'plain',
                      profanity_action: str = 'NoAction',
                      include_alignment: bool = False,
                      include_sentence_length: bool = False) -> List[Dict]:
        """
        Translate text to target language(s).
        
        Args:
            texts: String or list of strings to translate
            target_languages: Target language code(s)
            source_language: Source language code (optional, auto-detected if not provided)
            text_type: 'plain' or 'html'
            profanity_action: 'NoAction', 'Marked', or 'Deleted'
            include_alignment: Include word alignment information
            include_sentence_length: Include sentence length information
            
        Returns:
            List of translation results
        """
        if isinstance(texts, str):
            texts = [texts]
        if isinstance(target_languages, str):
            target_languages = [target_languages]
        
        # Build query parameters
        params = {
            'api-version': '3.0',
            'textType': text_type,
            'profanityAction': profanity_action
        }
        
        # Add target languages
        for lang in target_languages:
            params[f'to'] = lang
        
        # Add source language if specified
        if source_language:
            params['from'] = source_language
        
        # Add optional parameters
        if include_alignment:
            params['includeAlignment'] = 'true'
        if include_sentence_length:
            params['includeSentenceLength'] = 'true'
        
        body = [{'text': text} for text in texts]
        
        try:
            # Build URL with multiple 'to' parameters
            url = f"{self.endpoint}translate?api-version=3.0"
            for lang in target_languages:
                url += f"&to={lang}"
            
            if source_language:
                url += f"&from={source_language}"
            if text_type != 'plain':
                url += f"&textType={text_type}"
            if profanity_action != 'NoAction':
                url += f"&profanityAction={profanity_action}"
            if include_alignment:
                url += "&includeAlignment=true"
            if include_sentence_length:
                url += "&includeSentenceLength=true"
            
            response = self.session.post(url, json=body)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error translating text: {e}")
            return []
    
    def get_supported_languages(self, scope: str = 'translation') -> Dict:
        """
        Get list of supported languages.
        
        Args:
            scope: 'translation', 'transliteration', or 'dictionary'
            
        Returns:
            Dictionary of supported languages
        """
        try:
            response = self.session.get(
                f"{self.endpoint}languages?api-version=3.0&scope={scope}"
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error getting supported languages: {e}")
            return {}
    
    def transliterate_text(self, 
                          texts: Union[str, List[str]], 
                          language: str, 
                          from_script: str, 
                          to_script: str) -> List[Dict]:
        """
        Transliterate text from one script to another.
        
        Args:
            texts: String or list of strings to transliterate
            language: Language code
            from_script: Source script code
            to_script: Target script code
            
        Returns:
            List of transliteration results
        """
        if isinstance(texts, str):
            texts = [texts]
        
        body = [{'text': text} for text in texts]
        
        try:
            response = self.session.post(
                f"{self.endpoint}transliterate?api-version=3.0&language={language}&fromScript={from_script}&toScript={to_script}",
                json=body
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error transliterating text: {e}")
            return []

class TranslationApp:
    """Interactive translation application demonstrating various scenarios."""
    
    def __init__(self):
        """Initialize the translation app."""
        self.translator = AzureTranslator()
        self.translation_history = []
    
    def detect_and_display_language(self, text: str):
        """Detect and display language information for text."""
        print(f"\n🔍 Detecting language for: '{text[:50]}{'...' if len(text) > 50 else ''}'")
        
        results = self.translator.detect_language(text)
        if results:
            detection = results[0]
            language = detection.get('language', 'unknown')
            confidence = detection.get('score', 0)
            
            print(f"   Language: {language}")
            print(f"   Confidence: {confidence:.2%}")
            
            # Show alternatives if available
            if 'alternatives' in detection:
                print("   Alternatives:")
                for alt in detection['alternatives'][:3]:  # Show top 3
                    print(f"     - {alt['language']}: {alt['score']:.2%}")
            
            return language
        return None
    
    def translate_and_display(self, text: str, target_languages: List[str], source_language: Optional[str] = None):
        """Translate text and display results."""
        print(f"\n🌍 Translating to: {', '.join(target_languages)}")
        
        results = self.translator.translate_text(
            text, 
            target_languages, 
            source_language=source_language,
            include_alignment=True
        )
        
        if results:
            for result in results:
                detected_lang = result.get('detectedLanguage', {})
                if detected_lang:
                    print(f"   Detected: {detected_lang.get('language', 'unknown')} (confidence: {detected_lang.get('score', 0):.2%})")
                
                print(f"   Original: {text}")
                
                for translation in result.get('translations', []):
                    target_lang = translation.get('to', 'unknown')
                    translated_text = translation.get('text', '')
                    
                    print(f"   {target_lang.upper()}: {translated_text}")
                    
                    # Store in history
                    self.translation_history.append({
                        'original': text,
                        'translated': translated_text,
                        'source_lang': source_language or detected_lang.get('language'),
                        'target_lang': target_lang,
                        'timestamp': str(uuid.uuid4())[:8]
                    })
    
    def demonstrate_basic_translation(self):
        """Demonstrate basic text translation."""
        print("\n" + "="*50)
        print("🚀 BASIC TRANSLATION DEMO")
        print("="*50)
        
        # Single language translation
        text = "Hello, how are you today?"
        self.translate_and_display(text, ['es', 'fr', 'de'])
        
        # Multiple texts
        texts = [
            "Good morning!",
            "Thank you for your help.",
            "See you later!"
        ]
        
        print(f"\n📝 Translating multiple texts to Spanish:")
        for text in texts:
            results = self.translator.translate_text(text, 'es')
            if results:
                translation = results[0]['translations'][0]['text']
                print(f"   EN: {text}")
                print(f"   ES: {translation}")
    
    def demonstrate_language_detection(self):
        """Demonstrate language detection capabilities."""
        print("\n" + "="*50)
        print("🔍 LANGUAGE DETECTION DEMO")
        print("="*50)
        
        multilingual_texts = [
            "Hello, how are you?",
            "Bonjour, comment allez-vous?",
            "Hola, ¿cómo estás?",
            "Guten Tag, wie geht es Ihnen?",
            "こんにちは、元気ですか？",
            "Привет, как дела?",
            "مرحبا، كيف حالك؟"
        ]
        
        for text in multilingual_texts:
            self.detect_and_display_language(text)
    
    def demonstrate_html_translation(self):
        """Demonstrate HTML content translation."""
        print("\n" + "="*50)
        print("🌐 HTML TRANSLATION DEMO")
        print("="*50)
        
        html_content = """
        <h1>Welcome to our website!</h1>
        <p>We offer <strong>amazing products</strong> at great prices.</p>
        <p>Contact us at <a href="mailto:info@example.com">info@example.com</a></p>
        """
        
        print("Original HTML:")
        print(html_content)
        
        results = self.translator.translate_text(
            html_content, 
            ['es'], 
            text_type='html'
        )
        
        if results:
            translated_html = results[0]['translations'][0]['text']
            print("\nTranslated HTML (Spanish):")
            print(translated_html)
    
    def demonstrate_business_scenarios(self):
        """Demonstrate real-world business scenarios."""
        print("\n" + "="*50)
        print("💼 BUSINESS SCENARIOS DEMO")
        print("="*50)
        
        # Customer support scenario
        print("\n📞 Customer Support Scenario:")
        customer_message = "I'm having trouble with my order. Can you help me track it?"
        self.translate_and_display(customer_message, ['es', 'fr', 'de'])
        
        # Product description scenario
        print("\n🛍️ Product Description Scenario:")
        product_desc = "Premium wireless headphones with noise cancellation and 30-hour battery life."
        self.translate_and_display(product_desc, ['ja', 'ko', 'zh'])
        
        # Legal disclaimer scenario
        print("\n⚖️ Legal Disclaimer Scenario:")
        disclaimer = "By using this service, you agree to our terms and conditions."
        self.translate_and_display(disclaimer, ['pt', 'it', 'nl'])
    
    def demonstrate_error_handling(self):
        """Demonstrate error handling scenarios."""
        print("\n" + "="*50)
        print("⚠️ ERROR HANDLING DEMO")
        print("="*50)
        
        # Invalid language code
        print("\n❌ Testing invalid language code:")
        results = self.translator.translate_text("Hello", ['invalid_lang'])
        if not results:
            print("   Handled invalid language code gracefully")
        
        # Empty text
        print("\n❌ Testing empty text:")
        results = self.translator.translate_text("", ['es'])
        print(f"   Empty text result: {results}")
        
        # Very long text (testing limits)
        print("\n❌ Testing very long text:")
        long_text = "This is a test. " * 1000  # Create long text
        print(f"   Text length: {len(long_text)} characters")
        results = self.translator.translate_text(long_text[:5000], ['es'])  # Limit to 5000 chars
        if results:
            print("   Successfully handled long text")
    
    def show_translation_history(self):
        """Display translation history."""
        print("\n" + "="*50)
        print("📚 TRANSLATION HISTORY")
        print("="*50)
        
        if not self.translation_history:
            print("No translations in history.")
            return
        
        for i, translation in enumerate(self.translation_history[-5:], 1):  # Show last 5
            print(f"\n{i}. [{translation['timestamp']}]")
            print(f"   {translation['source_lang']} → {translation['target_lang']}")
            print(f"   Original: {translation['original'][:50]}{'...' if len(translation['original']) > 50 else ''}")
            print(f"   Translation: {translation['translated'][:50]}{'...' if len(translation['translated']) > 50 else ''}")
    
    def interactive_mode(self):
        """Run interactive translation mode."""
        print("\n" + "="*50)
        print("🎯 INTERACTIVE TRANSLATION MODE")
        print("="*50)
        print("Commands:")
        print("  - Type text to translate")
        print("  - 'lang:XX' to set target language (e.g., 'lang:es')")
        print("  - 'detect' to detect language of next input")
        print("  - 'history' to show translation history")
        print("  - 'quit' to exit")
        
        target_lang = 'es'  # Default to Spanish
        detect_mode = False
        
        while True:
            try:
                user_input = input(f"\n[Target: {target_lang}] Enter text: ").strip()
                
                if user_input.lower() == 'quit':
                    break
                elif user_input.lower() == 'history':
                    self.show_translation_history()
                elif user_input.lower() == 'detect':
                    detect_mode = True
                    print("Language detection mode enabled for next input.")
                elif user_input.startswith('lang:'):
                    target_lang = user_input[5:].strip()
                    print(f"Target language set to: {target_lang}")
                elif user_input:
                    if detect_mode:
                        self.detect_and_display_language(user_input)
                        detect_mode = False
                    else:
                        self.translate_and_display(user_input, [target_lang])
                
            except KeyboardInterrupt:
                print("\nExiting interactive mode...")
                break
            except Exception as e:
                print(f"Error: {e}")

def main():
    """Main function demonstrating Azure AI Translator capabilities."""
    print("🌍 Azure AI Translator - Text Translation Demo")
    print("=" * 60)
    
    try:
        # Initialize the translation app
        app = TranslationApp()
        
        # Check if we can connect to the service
        languages = app.translator.get_supported_languages()
        if languages:
            translation_langs = languages.get('translation', {})
            print(f"✅ Connected! Supports {len(translation_langs)} languages for translation")
        else:
            print("❌ Failed to connect to Azure Translator service")
            return
        
        # Run demonstrations
        app.demonstrate_basic_translation()
        app.demonstrate_language_detection()
        app.demonstrate_html_translation()
        app.demonstrate_business_scenarios()
        app.demonstrate_error_handling()
        app.show_translation_history()
        
        # Ask if user wants interactive mode
        print("\n" + "="*50)
        response = input("Would you like to try interactive mode? (y/n): ").strip().lower()
        if response in ['y', 'yes']:
            app.interactive_mode()
        
        print("\n🎉 Demo completed successfully!")
        print("\nKey takeaways:")
        print("✅ Text translation to multiple languages")
        print("✅ Automatic language detection")
        print("✅ HTML content translation")
        print("✅ Error handling and edge cases")
        print("✅ Production-ready patterns")
        
    except Exception as e:
        print(f"❌ Error running demo: {e}")
        print("\nPlease check your environment variables:")
        print("- TRANSLATOR_KEY")
        print("- TRANSLATOR_ENDPOINT") 
        print("- TRANSLATOR_REGION")

if __name__ == "__main__":
    main() 

File: Code-Samples/JavaScript/02-Text-and-Language/translate-text.js
#!/usr/bin/env node

/**
 * Azure AI Translator - Text Translation Example
 * 
 * This example demonstrates how to use Azure AI Translator service to:
 * - Detect language of input text
 * - Translate text to multiple target languages
 * - Handle various translation scenarios
 * - Implement best practices for production use
 * 
 * Prerequisites:
 * - Node.js 16+ installed
 * - npm install axios dotenv readline
 * 
 * Set environment variables:
 * - TRANSLATOR_KEY: Your Translator resource key
 * - TRANSLATOR_ENDPOINT: Your Translator resource endpoint
 * - TRANSLATOR_REGION: Your Translator resource region
 */

const axios = require('axios');
const readline = require('readline');
require('dotenv').config();

class AzureTranslator {
    /**
     * Azure AI Translator client for text translation operations.
     */
    constructor() {
        this.key = process.env.TRANSLATOR_KEY;
        this.endpoint = process.env.TRANSLATOR_ENDPOINT;
        this.region = process.env.TRANSLATOR_REGION;
        
        if (!this.key || !this.endpoint || !this.region) {
            throw new Error('Missing required environment variables. Please set TRANSLATOR_KEY, TRANSLATOR_ENDPOINT, and TRANSLATOR_REGION');
        }
        
        // Ensure endpoint ends with /
        if (!this.endpoint.endsWith('/')) {
            this.endpoint += '/';
        }
        
        // Create axios instance with default headers
        this.client = axios.create({
            baseURL: this.endpoint,
            headers: {
                'Ocp-Apim-Subscription-Key': this.key,
                'Ocp-Apim-Subscription-Region': this.region,
                'Content-Type': 'application/json',
                'X-ClientTraceId': this.generateUUID()
            },
            timeout: 30000 // 30 seconds timeout
        });
    }
    
    /**
     * Generate a UUID for request tracing.
     */
    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c === 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
    
    /**
     * Detect the language of input text(s).
     * 
     * @param {string|string[]} texts - String or array of strings to detect language for
     * @returns {Promise<Object[]>} Array of detection results
     */
    async detectLanguage(texts) {
        if (typeof texts === 'string') {
            texts = [texts];
        }
        
        const body = texts.map(text => ({ text }));
        
        try {
            const response = await this.client.post('detect?api-version=3.0', body);
            return response.data;
        } catch (error) {
            console.error('Error detecting language:', error.message);
            return [];
        }
    }
    
    /**
     * Translate text to target language(s).
     * 
     * @param {string|string[]} texts - String or array of strings to translate
     * @param {string|string[]} targetLanguages - Target language code(s)
     * @param {Object} options - Translation options
     * @returns {Promise<Object[]>} Array of translation results
     */
    async translateText(texts, targetLanguages, options = {}) {
        if (typeof texts === 'string') {
            texts = [texts];
        }
        if (typeof targetLanguages === 'string') {
            targetLanguages = [targetLanguages];
        }
        
        const {
            sourceLanguage = null,
            textType = 'plain',
            profanityAction = 'NoAction',
            includeAlignment = false,
            includeSentenceLength = false
        } = options;
        
        // Build query parameters
        const params = new URLSearchParams({
            'api-version': '3.0',
            textType,
            profanityAction
        });
        
        // Add target languages
        targetLanguages.forEach(lang => params.append('to', lang));
        
        // Add optional parameters
        if (sourceLanguage) params.set('from', sourceLanguage);
        if (includeAlignment) params.set('includeAlignment', 'true');
        if (includeSentenceLength) params.set('includeSentenceLength', 'true');
        
        const body = texts.map(text => ({ text }));
        
        try {
            const response = await this.client.post(`translate?${params}`, body);
            return response.data;
        } catch (error) {
            console.error('Error translating text:', error.message);
            return [];
        }
    }
    
    /**
     * Get list of supported languages.
     * 
     * @param {string} scope - 'translation', 'transliteration', or 'dictionary'
     * @returns {Promise<Object>} Dictionary of supported languages
     */
    async getSupportedLanguages(scope = 'translation') {
        try {
            const response = await this.client.get(`languages?api-version=3.0&scope=${scope}`);
            return response.data;
        } catch (error) {
            console.error('Error getting supported languages:', error.message);
            return {};
        }
    }
    
    /**
     * Transliterate text from one script to another.
     * 
     * @param {string|string[]} texts - String or array of strings to transliterate
     * @param {string} language - Language code
     * @param {string} fromScript - Source script code
     * @param {string} toScript - Target script code
     * @returns {Promise<Object[]>} Array of transliteration results
     */
    async transliterateText(texts, language, fromScript, toScript) {
        if (typeof texts === 'string') {
            texts = [texts];
        }
        
        const body = texts.map(text => ({ text }));
        const params = new URLSearchParams({
            'api-version': '3.0',
            language,
            fromScript,
            toScript
        });
        
        try {
            const response = await this.client.post(`transliterate?${params}`, body);
            return response.data;
        } catch (error) {
            console.error('Error transliterating text:', error.message);
            return [];
        }
    }
}

class TranslationApp {
    /**
     * Interactive translation application demonstrating various scenarios.
     */
    constructor() {
        this.translator = new AzureTranslator();
        this.translationHistory = [];
    }
    
    /**
     * Detect and display language information for text.
     */
    async detectAndDisplayLanguage(text) {
        console.log(`\n🔍 Detecting language for: '${text.substring(0, 50)}${text.length > 50 ? '...' : ''}'`);
        
        const results = await this.translator.detectLanguage(text);
        if (results.length > 0) {
            const detection = results[0];
            const language = detection.language || 'unknown';
            const confidence = detection.score || 0;
            
            console.log(`   Language: ${language}`);
            console.log(`   Confidence: ${(confidence * 100).toFixed(1)}%`);
            
            // Show alternatives if available
            if (detection.alternatives) {
                console.log('   Alternatives:');
                detection.alternatives.slice(0, 3).forEach(alt => {
                    console.log(`     - ${alt.language}: ${(alt.score * 100).toFixed(1)}%`);
                });
            }
            
            return language;
        }
        return null;
    }
    
    /**
     * Translate text and display results.
     */
    async translateAndDisplay(text, targetLanguages, sourceLanguage = null) {
        console.log(`\n🌍 Translating to: ${targetLanguages.join(', ')}`);
        
        const results = await this.translator.translateText(
            text, 
            targetLanguages, 
            { 
                sourceLanguage,
                includeAlignment: true 
            }
        );
        
        if (results.length > 0) {
            for (const result of results) {
                const detectedLang = result.detectedLanguage;
                if (detectedLang) {
                    console.log(`   Detected: ${detectedLang.language} (confidence: ${(detectedLang.score * 100).toFixed(1)}%)`);
                }
                
                console.log(`   Original: ${text}`);
                
                for (const translation of result.translations || []) {
                    const targetLang = translation.to || 'unknown';
                    const translatedText = translation.text || '';
                    
                    console.log(`   ${targetLang.toUpperCase()}: ${translatedText}`);
                    
                    // Store in history
                    this.translationHistory.push({
                        original: text,
                        translated: translatedText,
                        sourceLang: sourceLanguage || detectedLang?.language,
                        targetLang: targetLang,
                        timestamp: new Date().toISOString().substring(0, 19)
                    });
                }
            }
        }
    }
    
    /**
     * Demonstrate basic text translation.
     */
    async demonstrateBasicTranslation() {
        console.log('\n' + '='.repeat(50));
        console.log('🚀 BASIC TRANSLATION DEMO');
        console.log('='.repeat(50));
        
        // Single language translation
        const text = "Hello, how are you today?";
        await this.translateAndDisplay(text, ['es', 'fr', 'de']);
        
        // Multiple texts
        const texts = [
            "Good morning!",
            "Thank you for your help.",
            "See you later!"
        ];
        
        console.log('\n📝 Translating multiple texts to Spanish:');
        for (const text of texts) {
            const results = await this.translator.translateText(text, 'es');
            if (results.length > 0) {
                const translation = results[0].translations[0].text;
                console.log(`   EN: ${text}`);
                console.log(`   ES: ${translation}`);
            }
        }
    }
    
    /**
     * Demonstrate language detection capabilities.
     */
    async demonstrateLanguageDetection() {
        console.log('\n' + '='.repeat(50));
        console.log('🔍 LANGUAGE DETECTION DEMO');
        console.log('='.repeat(50));
        
        const multilingualTexts = [
            "Hello, how are you?",
            "Bonjour, comment allez-vous?",
            "Hola, ¿cómo estás?",
            "Guten Tag, wie geht es Ihnen?",
            "こんにちは、元気ですか？",
            "Привет, как дела?",
            "مرحبا، كيف حالك؟"
        ];
        
        for (const text of multilingualTexts) {
            await this.detectAndDisplayLanguage(text);
        }
    }
    
    /**
     * Demonstrate HTML content translation.
     */
    async demonstrateHtmlTranslation() {
        console.log('\n' + '='.repeat(50));
        console.log('🌐 HTML TRANSLATION DEMO');
        console.log('='.repeat(50));
        
        const htmlContent = `
        <h1>Welcome to our website!</h1>
        <p>We offer <strong>amazing products</strong> at great prices.</p>
        <p>Contact us at <a href="mailto:info@example.com">info@example.com</a></p>
        `;
        
        console.log('Original HTML:');
        console.log(htmlContent);
        
        const results = await this.translator.translateText(
            htmlContent, 
            ['es'], 
            { textType: 'html' }
        );
        
        if (results.length > 0) {
            const translatedHtml = results[0].translations[0].text;
            console.log('\nTranslated HTML (Spanish):');
            console.log(translatedHtml);
        }
    }
    
    /**
     * Demonstrate real-world business scenarios.
     */
    async demonstrateBusinessScenarios() {
        console.log('\n' + '='.repeat(50));
        console.log('💼 BUSINESS SCENARIOS DEMO');
        console.log('='.repeat(50));
        
        // Customer support scenario
        console.log('\n📞 Customer Support Scenario:');
        const customerMessage = "I'm having trouble with my order. Can you help me track it?";
        await this.translateAndDisplay(customerMessage, ['es', 'fr', 'de']);
        
        // Product description scenario
        console.log('\n🛍️ Product Description Scenario:');
        const productDesc = "Premium wireless headphones with noise cancellation and 30-hour battery life.";
        await this.translateAndDisplay(productDesc, ['ja', 'ko', 'zh']);
        
        // Legal disclaimer scenario
        console.log('\n⚖️ Legal Disclaimer Scenario:');
        const disclaimer = "By using this service, you agree to our terms and conditions.";
        await this.translateAndDisplay(disclaimer, ['pt', 'it', 'nl']);
    }
    
    /**
     * Demonstrate error handling scenarios.
     */
    async demonstrateErrorHandling() {
        console.log('\n' + '='.repeat(50));
        console.log('⚠️ ERROR HANDLING DEMO');
        console.log('='.repeat(50));
        
        // Invalid language code
        console.log('\n❌ Testing invalid language code:');
        const results1 = await this.translator.translateText("Hello", ['invalid_lang']);
        if (results1.length === 0) {
            console.log('   Handled invalid language code gracefully');
        }
        
        // Empty text
        console.log('\n❌ Testing empty text:');
        const results2 = await this.translator.translateText("", ['es']);
        console.log(`   Empty text result: ${JSON.stringify(results2)}`);
        
        // Very long text (testing limits)
        console.log('\n❌ Testing very long text:');
        const longText = "This is a test. ".repeat(1000);
        console.log(`   Text length: ${longText.length} characters`);
        const results3 = await this.translator.translateText(longText.substring(0, 5000), ['es']);
        if (results3.length > 0) {
            console.log('   Successfully handled long text');
        }
    }
    
    /**
     * Display translation history.
     */
    showTranslationHistory() {
        console.log('\n' + '='.repeat(50));
        console.log('📚 TRANSLATION HISTORY');
        console.log('='.repeat(50));
        
        if (this.translationHistory.length === 0) {
            console.log('No translations in history.');
            return;
        }
        
        const recentHistory = this.translationHistory.slice(-5); // Show last 5
        recentHistory.forEach((translation, index) => {
            console.log(`\n${index + 1}. [${translation.timestamp}]`);
            console.log(`   ${translation.sourceLang} → ${translation.targetLang}`);
            const originalPreview = translation.original.length > 50 
                ? translation.original.substring(0, 50) + '...' 
                : translation.original;
            const translatedPreview = translation.translated.length > 50 
                ? translation.translated.substring(0, 50) + '...' 
                : translation.translated;
            console.log(`   Original: ${originalPreview}`);
            console.log(`   Translation: ${translatedPreview}`);
        });
    }
    
    /**
     * Run interactive translation mode.
     */
    async interactiveMode() {
        console.log('\n' + '='.repeat(50));
        console.log('🎯 INTERACTIVE TRANSLATION MODE');
        console.log('='.repeat(50));
        console.log('Commands:');
        console.log('  - Type text to translate');
        console.log('  - "lang:XX" to set target language (e.g., "lang:es")');
        console.log('  - "detect" to detect language of next input');
        console.log('  - "history" to show translation history');
        console.log('  - "quit" to exit');
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        
        let targetLang = 'es'; // Default to Spanish
        let detectMode = false;
        
        const promptUser = () => {
            rl.question(`\n[Target: ${targetLang}] Enter text: `, async (userInput) => {
                const input = userInput.trim();
                
                if (input.toLowerCase() === 'quit') {
                    rl.close();
                    return;
                } else if (input.toLowerCase() === 'history') {
                    this.showTranslationHistory();
                    promptUser();
                } else if (input.toLowerCase() === 'detect') {
                    detectMode = true;
                    console.log('Language detection mode enabled for next input.');
                    promptUser();
                } else if (input.startsWith('lang:')) {
                    targetLang = input.substring(5).trim();
                    console.log(`Target language set to: ${targetLang}`);
                    promptUser();
                } else if (input) {
                    try {
                        if (detectMode) {
                            await this.detectAndDisplayLanguage(input);
                            detectMode = false;
                        } else {
                            await this.translateAndDisplay(input, [targetLang]);
                        }
                    } catch (error) {
                        console.error(`Error: ${error.message}`);
                    }
                    promptUser();
                } else {
                    promptUser();
                }
            });
        };
        
        promptUser();
    }
}

/**
 * Main function demonstrating Azure AI Translator capabilities.
 */
async function main() {
    console.log('🌍 Azure AI Translator - Text Translation Demo');
    console.log('='.repeat(60));
    
    try {
        // Initialize the translation app
        const app = new TranslationApp();
        
        // Check if we can connect to the service
        const languages = await app.translator.getSupportedLanguages();
        if (languages.translation) {
            const translationLangs = Object.keys(languages.translation);
            console.log(`✅ Connected! Supports ${translationLangs.length} languages for translation`);
        } else {
            console.log('❌ Failed to connect to Azure Translator service');
            return;
        }
        
        // Run demonstrations
        await app.demonstrateBasicTranslation();
        await app.demonstrateLanguageDetection();
        await app.demonstrateHtmlTranslation();
        await app.demonstrateBusinessScenarios();
        await app.demonstrateErrorHandling();
        app.showTranslationHistory();
        
        // Ask if user wants interactive mode
        console.log('\n' + '='.repeat(50));
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        
        rl.question('Would you like to try interactive mode? (y/n): ', (response) => {
            if (response.trim().toLowerCase() === 'y' || response.trim().toLowerCase() === 'yes') {
                rl.close();
                app.interactiveMode();
            } else {
                rl.close();
                console.log('\n🎉 Demo completed successfully!');
                console.log('\nKey takeaways:');
                console.log('✅ Text translation to multiple languages');
                console.log('✅ Automatic language detection');
                console.log('✅ HTML content translation');
                console.log('✅ Error handling and edge cases');
                console.log('✅ Production-ready patterns');
            }
        });
        
    } catch (error) {
        console.error(`❌ Error running demo: ${error.message}`);
        console.log('\nPlease check your environment variables:');
        console.log('- TRANSLATOR_KEY');
        console.log('- TRANSLATOR_ENDPOINT');
        console.log('- TRANSLATOR_REGION');
    }
}

// Run the demo if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { AzureTranslator, TranslationApp }; 